compiler/optimizer/VPHandlers.cpp

```diff
@@ -4241,13 +4241,65 @@ TR::Node *constrainNew(OMR::ValuePropagation *vp, TR::Node *node)
          vp->addGlobalConstraint(node, classConstraint->asClass()->getClassType());
       else
          vp->addGlobalConstraint(node, classConstraint);
+
+      // If it's an abstract class or an interface, the allocation cannot be removed
+      // because InstantiationError could be thrown
+      TR_OpaqueClassBlock *clazz = classConstraint->getClassType() ? classConstraint->getClassType()->getClass() : NULL;
+      if (clazz && TR::Compiler->cls.isConcreteClass(vp->comp(), clazz))
+         {
+         node->setAllocationCanBeRemoved(true);
+         }
       }
 
-   if (classConstraint && classConstraint->getClassType() &&
-      classConstraint->getClassType()->getClass())
+   vp->addGlobalConstraint(node, TR::VPNonNullObject::create(vp));
+
+   node->setIsNonNull(true);
+
+   return node;
+   }
+
+TR::Node *constrainNewvalue(OMR::ValuePropagation *vp, TR::Node *node)
+   {
+   constrainChildren(vp, node);
+
+   vp->createExceptionEdgeConstraints(TR::Block::CanCatchNewvalue, NULL, node);
+
+   bool trace = vp->trace();
+
+   // The constraints on the child can be applied to this node. We also know
+   // that the type is fixed and non-null.
+   //
+   bool isGlobal;
+   TR::VPConstraint *classConstraint = vp->getConstraint(node->getFirstChild(), isGlobal);
+   if (classConstraint)
       {
-      node->setAllocationCanBeRemoved(true);
+      // since loadaddrs are now primed with a ClassObject property,
+      // this property should not be propagated to the 'newvalue' unless the
+      // underlying type of the loadaddr is a java/lang/Class
+      //
+      // If the underlying type of the loadaddr is a java/lang/Class, the property can be propagated to 'new'
+      //
+      if (classConstraint->getClass() && !classConstraint->isFixedClass())
+         {
+         vp->addGlobalConstraint(node, TR::VPFixedClass::create(vp, classConstraint->getClass()));
+         }
+      else if (classConstraint->asClass() && classConstraint->asClass()->getClassType() &&
+               !(classConstraint->asClass()->getClassType()->isClassObject() == TR_yes))
+         {
+         vp->addGlobalConstraint(node, classConstraint->asClass()->getClassType());
+         }
+      else
+         {
+         vp->addGlobalConstraint(node, classConstraint);
+         }
+
+      // Primitive value type cannot be abstract class
+      if (classConstraint->getClassType() && classConstraint->getClassType()->getClass())
```

Comment:
I think this needs a check similar to the one that you're adding for `constrainNew`, above.  Machine-generated bytetcode might have an `aconst_init` operation, say, that specifies a non-value type class.  Executing the instruction should result in an `IncompatibleClassChangeError`, so the allocation cannot be removed in that case.

In addition, I think the "nestmate" requirement might need to be considered before calling `setAllocationCanBeRemoved(true)`

Reply:
>I think this needs a check similar to the one that you're adding for constrainNew
Good catch! I'll add `TR::Compiler->cls.isValueTypeClass(clazz)` in the condition check.

Reply:
>I think the "nestmate" requirement might need to be considered before calling setAllocationCanBeRemoved(true)

I think we need to check if the referenced value class belongs to the same nest as the current class before `setAllocationCanBeRemoved(true)`.  I'm looking through the code and haven't seen such API yet ...

Reply:
I think [isClassVisible](https://github.com/eclipse-openj9/openj9/blob/691f52a58dc1c78f60a094d442aaaca895e1142b/runtime/compiler/env/VMJ9.cpp#L7170) should take care of nestmate check

Reply:
Addressed in f2616d4. I also need to add `isClassVisible` to OMR::ClassEnv that requires to add the corresponding change in OpenJ9 too. I just created https://github.com/eclipse-openj9/openj9/pull/14885. Please review https://github.com/eclipse-openj9/openj9/pull/14885 along with this one.